src/drivers/file/server.rs

```diff
@@ -151,8 +170,140 @@ impl DriverExt for FileServerExt {
     }
 
     fn create_endpoint_from_url(&self, url: &url::Url) -> Option<Arc<dyn Driver>> {
-        let host = url.host_str().unwrap();
-        let path = std::fs::canonicalize(&host).expect("Failed to get absolute path");
-        return Some(Arc::new(FileServer::new(path)));
+        Some(Arc::new(FileServer::new(url.path().into()).build()))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{collections::BTreeMap, str::FromStr};
+
+    use anyhow::Result;
+    use mavlink::{error::ParserError, MavFrame};
+    use tokio::sync::RwLock;
+
+    use super::*;
+    #[tokio::test]
+    async fn read_all_messages() -> Result<()> {
+        let (sender, _receiver) = tokio::sync::broadcast::channel(1000000);
+
+        let messages_received_per_id =
+            Arc::new(RwLock::new(BTreeMap::<u32, Vec<(u64, Protocol)>>::new()));
+        let messages_received_cloned = messages_received_per_id.clone();
+
+        let tlog_file = PathBuf::from_str("tests/files/00025-2024-04-22_18-49-07.tlog").unwrap();
+
+        let driver = FileServer::new(tlog_file.clone())
+            .on_message(move |args: (u64, Protocol)| {
+                let messages_received = messages_received_cloned.clone();
+
+                async move {
+                    let message_id = args.1.message_id();
+
+                    let mut messages_received = messages_received.write().await;
+                    if let Some(samples) = messages_received.get_mut(&message_id) {
+                        samples.push(args);
+                    } else {
+                        messages_received.insert(message_id, Vec::from([args]));
+                    }
+
+                    Ok(())
+                }
+            })
+            .build();
+
+        let receiver_task_handle = tokio::spawn({
+            let sender = sender.clone();
+            async move { driver.run(sender).await }
+        });
+
+        // let file_v1_messages = 2; // TODO:  Add support for V1 messages
+        let file_v2_messages = 30437;
+        let file_messages = file_v2_messages;
+        let mut total_messages_read = 0;
+        let res = tokio::time::timeout(tokio::time::Duration::from_secs(1), async {
+            loop {
+                let messages_received_per_id = messages_received_per_id.read().await.clone();
+                total_messages_read = messages_received_per_id
+                    .values()
+                    .into_iter()
+                    .fold(0, |acc, samples| acc + samples.len());
+
+                if total_messages_read > file_messages {
+                    // panic!("Reading duplicates!");
+                } else if total_messages_read == file_messages {
+                    break;
```

Comment:
The panic above should be uncommented, I was playing with that and forgot about it. The idea is that if we read more messages than the `file_messages`, there is something wrong with our implementation, like reading a message twice. This happened at some point, and that's why I'm not comparing >=, for example.

Reply:
Yes, my point is that the breaks makes the panic unreachable.

Reply:
How? I can reach the panic when setting `file_messages = 20000`, for example.

Reply:
I thought that "total_messages_read" would be updated as it goes, so in theory, there is a chance where it reaches the expected number but in the next iteration it could be bigger.